core/src/block.rs

```diff
@@ -12,594 +11,826 @@
+            transaction_validator: &TransactionValidator,
+            wsv: WorldStateView,
+            latest_block: Option<HashOf<VersionedCommittedBlock>>,
+            block_height: u64,
+            current_topology: &Topology,
+        ) -> Result<VersionedCommittedBlock, (Self, eyre::Report)> {
+            Candidate(self.0.into_v1())
+                .validate::<IS_GENESIS>(
+                    transaction_validator,
+                    wsv,
+                    latest_block,
+                    block_height,
+                    current_topology,
+                )
+                .map(Into::into)
+                .map_err(|(Candidate(block), error)| (Candidate(block.into()), error))
+        }
+    }
+
+    impl Candidate<CommittedBlock> {
+        /// Calculate the hash of the current block.
+        /// [`CommittedBlock`] should have the same hash as [`ValidBlock`].
+        #[inline]
+        pub fn hash(&self) -> HashOf<Self> {
+            HashOf::new(&self.0.header).transmute()
+        }
+
+        /// Return signatures that are verified with the `hash` of this block, removing all other signatures.
+        #[inline]
+        pub fn retain_verified_signatures(&mut self) -> impl Iterator<Item = &SignatureOf<Self>> {
+            self.0
+                .signatures
+                .retain_verified_by_hash(self.hash().transmute())
+                .map(SignatureOf::transmute_ref::<Self>)
+        }
+
+        /// Validate transaction hashes, verify signatures and produce [`CommittedBlock`]
+        ///
+        /// # Errors
+        /// - If transaction hashes don't match the hashes stored in the block header
+        /// - If signatures verification fails
+        pub fn validate<const IS_GENESIS: bool>(
+            self,
+            transaction_validator: &TransactionValidator,
+            wsv: WorldStateView,
+            latest_block: Option<HashOf<VersionedCommittedBlock>>,
+            block_height: u64,
+            current_topology: &Topology,
+        ) -> Result<CommittedBlock, (Self, eyre::Report)> {
+            let block = Candidate::<SignedBlock>::from(self)
+                .validate::<IS_GENESIS>(transaction_validator, wsv, latest_block, block_height)
+                .map_err(|(Candidate(block), error)| {
+                    (
+                        Candidate(CommittedBlock {
+                            header: block.header,
+                            transactions: block.transactions,
+                            rejected_transactions: block.rejected_transactions,
+                            event_recommendations: block.event_recommendations,
+                            signatures: block.signatures.transmute(),
+                        }),
+                        error,
+                    )
+                })?;
+
+            if IS_GENESIS {
+                Ok(block.commit_unchecked())
+            } else {
+                block.commit(current_topology).map_err(|(block, error)| {
```

Comment:
I think that #3247 has accidentally removed verification that the committed block has required number of signatures

Reply:
I created issue for this yesterday, assign yourself if you will fix this in current PR. #3397